File: backend/app/models/indicator.py
"""Indicator-related Pydantic models."""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel


class MAType(str, Enum):
    """Moving average type."""

    SMA = "SMA"
    EMA = "EMA"


class MovingAverageResult(BaseModel):
    """Moving average calculation result."""

    period: int
    ma_type: MAType
    values: list[tuple[datetime, float | None]]  # (timestamp, value)
    current_value: float | None = None
    current_price: float | None = None
    distance_percent: float | None = None  # % distance from current price


class RSIResult(BaseModel):
    """RSI calculation result."""

    period: int
    values: list[tuple[datetime, float | None]]  # (timestamp, value)
    current_value: float | None = None
    is_overbought: bool = False
    is_oversold: bool = False


class MACDResult(BaseModel):
    """MACD calculation result."""

    fast_period: int
    slow_period: int
    signal_period: int
    macd_line: list[tuple[datetime, float | None]]
    signal_line: list[tuple[datetime, float | None]]
    histogram: list[tuple[datetime, float | None]]
    current_macd: float | None = None
    current_signal: float | None = None
    current_histogram: float | None = None


class IndicatorData(BaseModel):
    """Combined indicator data for a stock."""

    symbol: str
    ma_20w: MovingAverageResult | None = None
    ma_50w: MovingAverageResult | None = None
    ma_100w: MovingAverageResult | None = None
    ma_200w: MovingAverageResult | None = None
    rsi: RSIResult | None = None
    macd: MACDResult | None = None

File: frontend/src/types/indicator.ts
export type MAType = 'SMA' | 'EMA';

export interface MovingAverageResult {
  period: number;
  ma_type: MAType;
  values: [string, number | null][];
  current_value: number | null;
  current_price: number | null;
  distance_percent: number | null;
}

export interface RSIResult {
  period: number;
  values: [string, number | null][];
  current_value: number | null;
  is_overbought: boolean;
  is_oversold: boolean;
}

export interface MACDResult {
  fast_period: number;
  slow_period: number;
  signal_period: number;
  macd_line: [string, number | null][];
  signal_line: [string, number | null][];
  histogram: [string, number | null][];
  current_macd: number | null;
  current_signal: number | null;
  current_histogram: number | null;
}

export interface IndicatorData {
  symbol: string;
  ma_20w: MovingAverageResult | null;
  ma_50w: MovingAverageResult | null;
  ma_100w: MovingAverageResult | null;
  ma_200w: MovingAverageResult | null;
  rsi: RSIResult | null;
  macd: MACDResult | null;
}

export interface ChartData {
  symbol: string;
  timeframe: string;
  period: string;
  ohlcv: import('./stock').OHLCV[];
  indicators: {
    ma_20w?: MovingAverageResult;
    ma_50w?: MovingAverageResult;
    ma_100w?: MovingAverageResult;
    ma_200w?: MovingAverageResult;
    rsi?: RSIResult;
    macd?: MACDResult;
  };
}
